### ISSUE TITLE
[v4] indent expects the wrong indentation for expressions inside template literals

### ISSUE THREAD

**Issue by mike-marcacci** [eslint/eslint#8880]
**Tell us about your environment**

ESLint Version: 4.1.1
Node Version: 8.1.3
npm Version: yarn 0.27.5

**What parser (default, Babel-ESLint, etc.) are you using?**
default and babel-eslint

**Please show your full configuration:**
<details>
<summary>Configuration</summary>

```js
{
  "env": {
    "browser": false,
    "node": true,
    "jest": true
  },
  "extends" : [
    "standard",
    "plugin:flowtype/recommended"
  ],
  "parser": "babel-eslint",
  "rules": {
    "brace-style": [2, "stroustrup"],
    "comma-dangle": [1, "only-multiline"],
    "indent": [2, "tab", { "SwitchCase": 1 }],
    "no-multiple-empty-lines": [1, {"max": 4}],
    "no-tabs": [0],
    "padded-blocks": [0],
    "quotes": [2, "single"],
    "semi": [2, "always"],
    "strict": [2, "never"]
  },
  "plugins": [
    "flowtype"
  ]
}
```

</details>

**What did you do? Please include the actual source code causing the issue.**

```js
const fields = [
	'foo',
	() => 'bar'
];

const query = `
	SELECT
		${fields.map((f) => typeof condition === 'function'
			? f()
			: f
		).join(',')}
	FROM foo
`;

console.log(query);
```

**What did you expect to happen?**
No errors.

**What actually happened? Please include the actual, raw output from ESLint.**
```
   9:1  error  Expected indentation of 2 tabs but found 3  indent
  10:1  error  Expected indentation of 2 tabs but found 3  indent
  11:1  error  Expected indentation of 1 tab but found 2   indent
```

**Comment by not-an-aardvark** [comment]
Thanks for the report.

If I'm understanding correctly, you expected the line with `? f()` to be indented by 3 tabs because the previous line has two tabs at the start. I think the source of confusion is that the two tabs on the previous line are not "indentation" in the strictest sense of the word -- they're actually a part of your template literal. (It looks like your template literal is a SQL query, so whitespace in the string would be semantically insignificant in your case, but we can't assume that in general about all strings.) The rule has to account for the presence of arbitrary text in the template literal, so it generally doesn't take into account whether the previous line in the template literal starts with tabs or spaces. Instead, it offsets template element contents by an indentation level from the left side of the file, if the `${` started on a line after the start of the template literal.

**Comment by mike-marcacci** [comment]
Ah yes, this makes sense. I'll have to revisit how I format these, since I agree that programmatically interpreting what I'm after in a way that is generic to all template literals is impractical.

Thanks for the detailed explanation!

**Comment by mike-marcacci** [comment]
Actually... as I've been going through my code to fix these, I've found a few other instances of non-SQL template literals which require similarly unergonomic spacing (but would follow different rules than I've described above).

Would you be generally amenable to the addition of an "ignore-indentation-in-template-literals" to `indent`? If so, I can write up a proper proposal.

**Comment by not-an-aardvark** [comment]
I think that would be covered by https://github.com/eslint/eslint/issues/8594.

**Comment by mike-marcacci** [comment]
Oh yes! That is a fantastic proposal – subscribed. Thanks again!

**Comment by skylize** [comment]
> I think that would be covered by #8594.

@not-an-aardvark Would you mind giving an example how to use this rule specifically for dealing with template literals?
